File: backend/app/services/llm.py
import os
from openai import OpenAI
import json

class LLMService:
    def __init__(self):
        # Initialize Groq client
        # User must set GROQ_API_KEY env variable
        api_key = os.getenv("GROQ_API_KEY")
        if not api_key:
             # Fallback or stub for development if key is missing
             print("WARNING: GROQ_API_KEY not found. LLM features will be disabled/mocked.")
             self.client = None
        else:
             self.client = OpenAI(
                 base_url="https://api.groq.com/openai/v1",
                 api_key=api_key
             )
        
        self.model = "llama-3.3-70b-versatile"

    def generate_text(self, prompt: str, system_prompt: str = "You are a helpful AI career coach.") -> str:
        if not self.client:
            return "LLM Service Unavailable (Missing API Key)"
        
        try:
            response = self.client.chat.completions.create(
                model=self.model,
                messages=[
                    {"role": "system", "content": system_prompt},
                    {"role": "user", "content": prompt}
                ],
                temperature=0.7,
                max_tokens=1024
            )
            return response.choices[0].message.content
        except Exception as e:
            print(f"LLM Error: {e}")
            return "Error generating content."

    def generate_json(self, prompt: str, system_prompt: str = "You are an AI that outputs JSON.") -> dict:
        """
        Forces JSON output from the LLM.
        """
        if not self.client:
            return {}
            
        try:
            response = self.client.chat.completions.create(
                model=self.model,
                messages=[
                    {"role": "system", "content": system_prompt + " Output valid JSON only. Do not wrap in markdown code blocks."},
                    {"role": "user", "content": prompt}
                ],
                response_format={"type": "json_object"},
                temperature=0.7
            )
            content = response.choices[0].message.content
            # Clean possible markdown
            content = content.replace("```json", "").replace("```", "").strip()
            return json.loads(content)
        except Exception as e:
            print(f"LLM JSON Error: {e}")
            print(f"Raw Content causing error: {content if 'content' in locals() else 'No content'}")
            return {}

File: app/api/lib/llm.ts
import OpenAI from 'openai';

// LLM Service using Groq API (OpenAI-compatible)
class LLMService {
    private client: OpenAI | null = null;
    private model = 'llama-3.3-70b-versatile';

    constructor() {
        const apiKey = process.env.GROQ_API_KEY;
        if (!apiKey) {
            console.warn('WARNING: GROQ_API_KEY not found. LLM features will be disabled/mocked.');
        } else {
            this.client = new OpenAI({
                baseURL: 'https://api.groq.com/openai/v1',
                apiKey: apiKey,
            });
        }
    }

    isAvailable(): boolean {
        return this.client !== null;
    }

    async generateText(
        prompt: string,
        systemPrompt: string = 'You are a helpful AI career coach.'
    ): Promise<string> {
        if (!this.client) {
            return 'LLM Service Unavailable (Missing API Key)';
        }

        try {
            const response = await this.client.chat.completions.create({
                model: this.model,
                messages: [
                    { role: 'system', content: systemPrompt },
                    { role: 'user', content: prompt },
                ],
                temperature: 0.7,
                max_tokens: 1024,
            });
            return response.choices[0]?.message?.content || '';
        } catch (error) {
            console.error('LLM Error:', error);
            return 'Error generating content.';
        }
    }

    async generateJSON<T = Record<string, unknown>>(
        prompt: string,
        systemPrompt: string = 'You are an AI that outputs JSON.'
    ): Promise<T> {
        if (!this.client) {
            return {} as T;
        }

        try {
            const response = await this.client.chat.completions.create({
                model: this.model,
                messages: [
                    {
                        role: 'system',
                        content: systemPrompt + ' Output valid JSON only. Do not wrap in markdown code blocks.',
                    },
                    { role: 'user', content: prompt },
                ],
                response_format: { type: 'json_object' },
                temperature: 0.7,
            });

            let content = response.choices[0]?.message?.content || '{}';
            // Clean possible markdown
            content = content.replace(/```json/g, '').replace(/```/g, '').trim();
            return JSON.parse(content) as T;
        } catch (error) {
            console.error('LLM JSON Error:', error);
            return {} as T;
        }
    }
}

// Export singleton instance
export const llmService = new LLMService();
